jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeouts.java

```diff
@@ -0,0 +1,167 @@
+ * @see CyclicTimeout
+ */
+public abstract class CyclicTimeouts<T extends CyclicTimeouts.Expirable> implements Destroyable
+{
+    private static final Logger LOG = Log.getLogger(CyclicTimeouts.class);
+
+    private final AtomicLong earliestTimeout = new AtomicLong(Long.MAX_VALUE);
+    private final CyclicTimeout cyclicTimeout;
+
+    public CyclicTimeouts(Scheduler scheduler)
+    {
+        cyclicTimeout = new CyclicTimeout(scheduler)
+        {
+            @Override
+            public void onTimeoutExpired()
+            {
+                CyclicTimeouts.this.onTimeoutExpired();
+            }
+        };
+    }
+
+    /**
+     * @return the entities to iterate over when this instance expires
+     */
+    protected abstract Iterator<T> iterator();
+
+    /**
+     * <p>Invoked during the iteration when the given entity is expired.</p>
+     *
+     * @param expirable the entity that is expired
+     * @return whether the entity should be removed from the iterator via {@link Iterator#remove()}
+     */
+    protected abstract boolean onExpired(T expirable);
+
+    private void onTimeoutExpired()
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} timeouts check", this);
+
+        long now = System.nanoTime();
+        long earliest = Long.MAX_VALUE;
+        // Reset the earliest timeout so we can expire again.
+        // A concurrent call to schedule(long) may lose an
+        // earliest value, but the corresponding entity will
+        // be seen during the iteration below.
+        earliestTimeout.set(earliest);
+
+        // Scan the entities to abort expired entities
+        // and to find the entity that expires the earliest.
+        Iterator<T> iterator = iterator();
+        while (iterator.hasNext())
+        {
+            T expirable = iterator.next();
+            long expiresAt = expirable.getExpireNanoTime();
+            if (expiresAt == -1)
+                continue;
+            if (expiresAt <= now)
+            {
+                if (onExpired(expirable))
+                    iterator.remove();
+                continue;
+            }
+            earliest = Math.min(earliest, expiresAt);
+        }
+
+        if (earliest < Long.MAX_VALUE)
+            schedule(earliest);
+    }
+
+    /**
+     * <p>Manages the timeout of a new entity.</p>
+     *
+     * @param expirable the new entity to manage the timeout for
+     */
+    public void schedule(T expirable)
+    {
+        long expiresAt = expirable.getExpireNanoTime();
+        if (expiresAt < Long.MAX_VALUE)
+            schedule(expiresAt);
+    }
+
+    private void schedule(long expiresAt)
+    {
+        // Schedule a timeout for the earliest entity that may expire.
+        // When the timeout expires, scan the entities for the next
+        // earliest entity that may expire, and reschedule a new timeout.
+        long prevEarliest = earliestTimeout.getAndUpdate(t -> Math.min(t, expiresAt));
```

Comment:
Should we check prevEarliest for a flag value that indicates `onTimeoutExpired()` is currently running and thus we should not do anything else here?